### PR TITLE
Added blind score requirement and reward in blind select screen

### PR DESCRIPTION
**Pull request by MeirGavish**
I added score requirement and reward display in the blind select screen. I used the score truncation to squeeze them into 3 tiles there.
I also did a small refactor to rename the `blinds` array into `blind_states` to make it a little clearer.

https://github.com/user-attachments/assets/5be6c95a-c973-4cb7-be56-4fe11eea9e48

Since the truncation is so extreme and rounded down, it's sometimes a little awkward like on ante 2 where it looks like both the Big Blind and Boss Blind have the same score requirement of 1K when in fact it's 1200 and 1600.

image

Another screenshot of the infamous ante 5
image

And on ante 8 there is inevitable overflow but it's handled as well as possible.

https://github.com/user-attachments/assets/08073de4-1400-48a7-82e6-f12313e9f346

### CODE REVIEW

**Comment by ricfehr3** [comment]
@MeirGavish Big thank you on this one. It looks nice. I'm going to give this a review later tonight. I wanted to propose something before I review. The lack of making `1.2K` for example bugs me a bit. I thought of a (__maybe unrealistic__) idea. Loading a custom font actually isn't too difficult. We could have a copy of the standard font, then map unused characters to custom, squished versions of the numbers text with decimals: "1.","2.",..."9.","0." or the decimal first ".0",".1"... etc. So, for example, we could map the following (Looking at the [tonc tutorial](https://gbadev.net/tonc/tte.html#ssec-test-se4), I don't think we use these characters):

1. '&' <-> '.0'
2. '^' <-> '.1'
3. '{' <-> '.2'
4. '}' <-> '.3'
5. '|' <-> '.4'
6. '`' <-> '.5'
7. '<' <-> '.6'
8. '>' <-> '.7'
9. '_' <-> '.8'
10.';' <-> '.9'

I have an [example of modifying the font](https://github.com/ricfehr3/balatro-gba/commit/faa4948daa297726be5740ee233f1a7619e4a5bf). It's a commit branched from this branch right now, so if you want to apply as a [patch](https://github.com/ricfehr3/balatro-gba/commit/faa4948daa297726be5740ee233f1a7619e4a5bf.patch). 

This would require:

- [ ] Artwork for all the numbers.
- [ ] Some code mapping the characters.  

Let me know what you think. I'll come back to give this a real review.

**Comment by MeirGavish** [comment]
> I wanted to propose something before I review. The lack of making `1.2K` for example bugs me a bit. I thought of a (**maybe unrealistic**) idea. Loading a custom font actually isn't too difficult. We could have a copy of the standard font, then map unused characters to custom, squished versions of the numbers text with decimals: "1.","2.",..."9.","0." or the decimal first ".0",".1"... etc. So, for example, we could map the following (Looking at the [tonc tutorial](https://gbadev.net/tonc/tte.html#ssec-test-se4), I don't think we use these characters):
> 
> 1. '&' <-> '.0'
> 2. '^' <-> '.1'
> 3. '{' <-> '.2'
> 4. '}' <-> '.3'
> 5. '|' <-> '.4'
> 6. '`' <-> '.5'
> 7. '<' <-> '.6'
> 8. '>' <-> '.7'
> 9. '_' <-> '.8'
>    10.';' <-> '.9'
> 
> I have an [example of modifying the font](https://github.com/ricfehr3/balatro-gba/commit/faa4948daa297726be5740ee233f1a7619e4a5bf). It's a commit branched from this branch right now, so if you want to apply as a [patch](https://github.com/ricfehr3/balatro-gba/commit/faa4948daa297726be5740ee233f1a7619e4a5bf.patch).
> 
> This would require:
> 
> * [ ]  Artwork for all the numbers.
> * [ ]  Some code mapping the characters.
> 
> Let me know what you think. I'll come back to give this a real review.

@ricfehr3 This looks like a cool idea! It might work but it means adding more logic around `snprintf()` in `truncate_uint_to_suffixed_str()` to handle the decimal point which might not be a big deal but I'm noting that. Does your example include the new character tiles? I suppose we need to look into making them and that they all work.

**Comment by ricfehr3** [comment]
> > I wanted to propose something before I review. The lack of making `1.2K` for example bugs me a bit. I thought of a (**maybe unrealistic**) idea. Loading a custom font actually isn't too difficult. We could have a copy of the standard font, then map unused characters to custom, squished versions of the numbers text with decimals: "1.","2.",..."9.","0." or the decimal first ".0",".1"... etc. So, for example, we could map the following (Looking at the [tonc tutorial](https://gbadev.net/tonc/tte.html#ssec-test-se4), I don't think we use these characters):
> > 
> > 1. '&' <-> '.0'
> > 2. '^' <-> '.1'
> > 3. '{' <-> '.2'
> > 4. '}' <-> '.3'
> > 5. '|' <-> '.4'
> > 6. '`' <-> '.5'
> > 7. '<' <-> '.6'
> > 8. '>' <-> '.7'
> > 9. '_' <-> '.8'
> >    10.';' <-> '.9'
> > 
> > I have an [example of modifying the font](https://github.com/ricfehr3/balatro-gba/commit/faa4948daa297726be5740ee233f1a7619e4a5bf). It's a commit branched from this branch right now, so if you want to apply as a [patch](https://github.com/ricfehr3/balatro-gba/commit/faa4948daa297726be5740ee233f1a7619e4a5bf.patch).
> > This would require:
> > 
> > * [ ]   Artwork for all the numbers.
> > * [ ]   Some code mapping the characters.
> > 
> > Let me know what you think. I'll come back to give this a real review.
> 
> @ricfehr3 This looks like a cool idea! It might work but it means adding more logic around `snprintf()` in `truncate_uint_to_suffixed_str()` to handle the decimal point which might not be a big deal but I'm noting that. Does your example include the new character tiles? I suppose we need to look into making them and that they all work.

Sweet! The font example I sent was very bare bones, I just replaced all the characters with empty spaces. There are some tools to generate these files. I'll give it a try and push it to that branch again.

**Comment by MeirGavish** [comment]
I also tested giving one more tile to the text and allow it to overflow and I think it doesn't look very good.
image

@ricfehr3 I might be able to look into the custom font, what tools did you use to generate it?

Edit: I opened issue #295 for this so this task can be assigned to whoever is working on it.
I also opened issue #294 for this PR, if you think it shouldn't be merged as is, we can make #295 blocking.

**Comment by ricfehr3** [comment]
Unblocked now as of #314

**Comment by MeirGavish** [comment]
Rebased on #333, it's beautiful!

image

**Comment by ricfehr3** [comment]
> Rebased on #333, it's beautiful!
> image

It really is a wonderful sight.

**Comment by MeirGavish** [comment]
Rebasing was a mistake... I'll take care of it...

**Comment by MeirGavish** [comment]
There, properly rebased on main, wasn't too bad.

File: source/game.c [hunk 1]
```diff
@@ -4362,6 +4364,101 @@ static void game_blind_select_on_update()
     blind_select_state_actions[substate]();
 }
 
+static void game_blind_select_erase_blind_reqs_and_rewards()
```
Comment (ricfehr3) [inline source/game.c]: Since this is only called in one place I think it makes sense to inline it.

File: source/game.c [hunk 2]
```diff
@@ -4362,6 +4364,101 @@ static void game_blind_select_on_update()
...
+        blind_req_and_reward_rect.left +=
+            curr_blind * rect_width(&SINGLE_BLIND_SELECT_RECT) * TILE_SIZE;
+        blind_req_and_reward_rect.right +=
+            curr_blind * rect_width(&SINGLE_BLIND_SELECT_RECT) * TILE_SIZE;
+
+        tte_erase_rect_wrapper(blind_req_and_reward_rect);
+    }
+}
+
+static Rect game_blind_select_get_req_score_rect(enum BlindType blind)
```
Comment (ricfehr3) [inline source/game.c:4391]: Since this is used in a couple places, it shouldn't be inlined, but it should be forward declared with other static functions for consistency.

File: source/game.c [hunk 3]
```diff
@@ -4362,6 +4364,101 @@ static void game_blind_select_on_update()
...
+    tte_printf(
+        "#{P:%d,%d; cx:0x%X000}%s",
+        blind_reward_rect.left,
+        blind_reward_rect.top,
+        TTE_YELLOW_PB,
+        blind_reward_str_buff
+    );
+}
+
+static void game_blind_select_print_blinds_reqs_and_rewards()
```
Comment (ricfehr3) [inline source/game.c]: Same as the [above comment](https://github.com/GBALATRO/balatro-gba/pull/282/changes#r2663612817), this should be forward declared with the other static functions.

**Review by ricfehr3 — CHANGES_REQUESTED** [review]
Great work @MeirGavish ! The changes here look really good on screen. I've requested some small changes for consistency in how static functions are declared, but otherwise, it should be good to go. Whenever you feel comfortable merging, go for it.

**Comment by MeirGavish** [comment]
@ricfehr3 Thanks! I updated the PR and re-requested a review in case you still have time to formally approve. Otherwise I will merge this later today.
